src/hotspot/share/memory/allocation.hpp

```diff
@@ -174,48 +174,137 @@ char* ReallocateHeap(char *old,
+    return (void*)AllocateHeap(size, f, stack);
+  }
+
+  static ALWAYSINLINE void* operator new[](size_t size, const std::nothrow_t&,
+                                    const NativeCallStack& stack, MEMFLAGS f) throw() {
+    return (void*)AllocateHeap(size, f, stack, AllocFailStrategy::RETURN_NULL);
+  }
+
+  static ALWAYSINLINE void* operator new[](size_t size, const std::nothrow_t&, MEMFLAGS f) throw() {
+    return (void*)AllocateHeap(size, f, AllocFailStrategy::RETURN_NULL);
+  }
+
+  static void  operator delete(void* p)     { FreeHeap(p); }
+  static void  operator delete [] (void* p) { FreeHeap(p); }
+};
+
+template<MEMFLAGS F>
+class CHeapObj {
  public:
   ALWAYSINLINE void* operator new(size_t size) throw() {
-    return (void*)AllocateHeap(size, F);
+    return CHeapObjImpl::operator new(size, F);
   }
 
   ALWAYSINLINE void* operator new(size_t size,
                                   const NativeCallStack& stack) throw() {
-    return (void*)AllocateHeap(size, F, stack);
+    return CHeapObjImpl::operator new(size, stack, F);
   }
 
-  ALWAYSINLINE void* operator new(size_t size, const std::nothrow_t&,
+  ALWAYSINLINE void* operator new(size_t size, const std::nothrow_t& nt,
                                   const NativeCallStack& stack) throw() {
-    return (void*)AllocateHeap(size, F, stack, AllocFailStrategy::RETURN_NULL);
+    return CHeapObjImpl::operator new(size, nt, stack, F);
   }
 
-  ALWAYSINLINE void* operator new(size_t size, const std::nothrow_t&) throw() {
-    return (void*)AllocateHeap(size, F, AllocFailStrategy::RETURN_NULL);
+  ALWAYSINLINE void* operator new(size_t size, const std::nothrow_t& nt) throw() {
+    return CHeapObjImpl::operator new(size, nt, F);
   }
 
   ALWAYSINLINE void* operator new[](size_t size) throw() {
-    return (void*)AllocateHeap(size, F);
+    return CHeapObjImpl::operator new[](size, F);
   }
 
   ALWAYSINLINE void* operator new[](size_t size,
-                                  const NativeCallStack& stack) throw() {
-    return (void*)AllocateHeap(size, F, stack);
+                                    const NativeCallStack& stack) throw() {
+    return CHeapObjImpl::operator new[](size, stack, F);
   }
 
-  ALWAYSINLINE void* operator new[](size_t size, const std::nothrow_t&,
+  ALWAYSINLINE void* operator new[](size_t size, const std::nothrow_t& nt,
                                     const NativeCallStack& stack) throw() {
-    return (void*)AllocateHeap(size, F, stack, AllocFailStrategy::RETURN_NULL);
+    return CHeapObjImpl::operator new[](size, nt, stack, F);
+  }
+
+  ALWAYSINLINE void* operator new[](size_t size, const std::nothrow_t& nt) throw() {
+    return CHeapObjImpl::operator new[](size, nt, F);
+  }
+
+  void  operator delete(void* p)     { CHeapObjImpl::operator delete(p); }
+  void  operator delete [] (void* p) { CHeapObjImpl::operator delete[](p); }
+
+};
+
+// Dynamically pick the memory flags at allocation
+class CHeapObjDynamic {
```

Comment:
This can be simply using CHeapObjDynamic = CHeapObjImpl; so not to repeat all the various new operator declarations.